missioncontrol/api/views.py

```diff
@@ -83,39 +92,41 @@ def measure(request):
 
     if not all([channel_name, platform_name, measure_name, interval]):
         return HttpResponseBadRequest("All of channel, platform, measure, interval required")
-    data = cache.get(get_measure_cache_key(platform_name, channel_name, measure_name))
-    if not data:
+
+    datums = Datum.objects.filter(
+        series__build__channel__name=channel_name,
+        series__build__platform__name=platform_name,
+        series__measure__name=measure_name)
+
+    if not datums.exists():
         return HttpResponseNotFound("Data not available for this measure combination")
 
     # process min/max time based on parameters
     try:
         if start is not None:
-            min_time = datetime.datetime.fromtimestamp(int(start))
-            max_time = min_time + datetime.timedelta(seconds=int(interval))
+            min_time = datetime.datetime.fromtimestamp(int(start), tz=pytz.UTC)
+            datums = datums.filter(
+                timestamp__gt=min_time,
+                timestamp__lt=min_time + datetime.timedelta(seconds=int(interval)))
         else:
-            (min_time, max_time) = (datetime.datetime.now() -
-                                    datetime.timedelta(seconds=int(interval)),
-                                    None)
+            datums = datums.filter(
+                timestamp__gt=(timezone.now() -
+                               datetime.timedelta(seconds=int(interval))))
     except ValueError:
         raise HttpResponseBadRequest(
             "Interval / start time must be specified in seconds (as an integer)")
 
-    empty_buildids = set()
-    for (build_id, build_data) in data.items():
-        datums = build_data['data']
-        # filter out elements that don't match our date range
-        if min_time:
-            datums = filter(lambda d: d[0] >= min_time, datums)
-        if max_time:
-            datums = filter(lambda d: d[0] <= max_time, datums)
-        datums = list(map(lambda d: [_dt_to_utc(d[0])] +
-                          list(d[1:]), datums))
-        if not datums:
-            empty_buildids.add(build_id)
-        build_data['data'] = datums
-
-    # don't bother returning empty indexed data
-    for empty_buildid in empty_buildids:
-        del data[empty_buildid]
+    data = {}
+    for (build_id, version) in datums.values_list(
+            'series__build__build_id',
+            'series__build__version').distinct():
+        data[build_id] = {
+            'data': [],
+            'version': version
+        }
+    for (build_id, timestamp, value, usage_hours) in datums.values_list(
+            'series__build__build_id',
+            'timestamp', 'value', 'usage_hours').order_by('timestamp'):
+        data[build_id]['data'].append((timestamp, value, usage_hours))
```

Comment:
There are a few joins here in the various queries. Have you checked the SQL being produced? I'd be curious if it may benefit from using `select_related` in some cases.

Reply:
I have, yeah. Since we're using values_list here (and not ORM objects), everything maps nicely down to a single SQL query. E.g.:

```
>>> datums = Datum.objects.filter(series__build__channel__name='release', series__build__platform__name='linux', series__measure__name='main_crashes')
>>> print(datums.values_list('series__build__build_id', 'timestamp', 'value', 'usage_hours').order_by('timestamp').query)
SELECT "build"."build_id", "datum"."timestamp", "datum"."value", "datum"."usage_hours" FROM "datum" INNER JOIN "series" ON ("datum"."series_id" = "series"."id") INNER JOIN "build" ON ("series"."build_id" = "build"."id") INNER JOIN "channel" ON ("build"."channel_id" = "channel"."id") INNER JOIN "platform" ON ("build"."platform_id" = "platform"."id") INNER JOIN "measure" ON ("series"."measure_id" = "measure"."id") WHERE ("channel"."name" = release AND "platform"."name" = linux AND "measure"."name" = main_crashes) ORDER BY "datum"."timestamp" ASC
```